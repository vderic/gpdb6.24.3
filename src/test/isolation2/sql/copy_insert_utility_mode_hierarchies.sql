-- We should disallow COPY FROM and INSERT in utility mode if the target table
-- is a root of a partition/inheritance hierarchy.

CREATE TABLE utility_part_root (a int, b int, c int)
    DISTRIBUTED BY (a)
    PARTITION BY range(b)
        SUBPARTITION BY range(c)
            SUBPARTITION TEMPLATE (
            START(40) END(46) EVERY(3)
            )
        (START(0) END(4) EVERY(2));

-- should complain as we are trying copy on a root or subroot
1U: COPY utility_part_root FROM '/dev/null';
1U: COPY utility_part_root_1_prt_1 FROM '/dev/null';

-- should be fine as we are trying it on a leaf
1U: COPY utility_part_root_1_prt_1_2_prt_1 FROM '/dev/null';

-- should complain as we are trying insert on a root or subroot
1U: INSERT INTO utility_part_root VALUES(1, 2, 41);
1U: INSERT INTO utility_part_root_1_prt_1 VALUES(1,0,40);

-- should be fine as we are trying it on a leaf
1U: INSERT INTO utility_part_root_1_prt_1_2_prt_1 VALUES(1,0,40);

CREATE TABLE utility_inh_root (
    a int
);
CREATE TABLE utility_inh_subroot (
    b int
) INHERITS (utility_inh_root);
CREATE TABLE utility_inh_leaf (
    c int
) INHERITS (utility_inh_subroot);

-- should complain as we are trying copy on a root or subroot
1U: COPY utility_inh_root FROM '/dev/null';
1U: COPY utility_inh_subroot FROM '/dev/null';

-- should be fine as we are trying it on a leaf
1U: COPY utility_inh_leaf FROM '/dev/null';

-- should complain as we are trying insert on a root or subroot
1U: INSERT INTO utility_inh_root VALUES(1);
1U: INSERT INTO utility_inh_subroot VALUES(1, 2);

-- should be fine as we are trying it on a leaf
1U: INSERT INTO utility_inh_leaf VALUES(1, 2, 3);
