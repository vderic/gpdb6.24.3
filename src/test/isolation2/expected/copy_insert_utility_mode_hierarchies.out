-- We should disallow COPY FROM and INSERT in utility mode if the target table
-- is a root of a partition/inheritance hierarchy.

CREATE TABLE utility_part_root (a int, b int, c int) DISTRIBUTED BY (a) PARTITION BY range(b) SUBPARTITION BY range(c) SUBPARTITION TEMPLATE ( START(40) END(46) EVERY(3) ) (START(0) END(4) EVERY(2));
CREATE

-- should complain as we are trying copy on a root or subroot
1U: COPY utility_part_root FROM '/dev/null';
ERROR:  COPY FROM in utility mode is not supported on roots of inheritance and partition hierarchies
HINT:  consider using COPY FROM with the ON SEGMENT clause
1U: COPY utility_part_root_1_prt_1 FROM '/dev/null';
ERROR:  COPY FROM in utility mode is not supported on roots of inheritance and partition hierarchies
HINT:  consider using COPY FROM with the ON SEGMENT clause

-- should be fine as we are trying it on a leaf
1U: COPY utility_part_root_1_prt_1_2_prt_1 FROM '/dev/null';
COPY 0

-- should complain as we are trying insert on a root or subroot
1U: INSERT INTO utility_part_root VALUES(1, 2, 41);
ERROR:  INSERT in utility mode is not supported on roots of inheritance and partition hierarchies
1U: INSERT INTO utility_part_root_1_prt_1 VALUES(1,0,40);
ERROR:  INSERT in utility mode is not supported on roots of inheritance and partition hierarchies

-- should be fine as we are trying it on a leaf
1U: INSERT INTO utility_part_root_1_prt_1_2_prt_1 VALUES(1,0,40);
INSERT 1

CREATE TABLE utility_inh_root ( a int );
CREATE
CREATE TABLE utility_inh_subroot ( b int ) INHERITS (utility_inh_root);
CREATE
CREATE TABLE utility_inh_leaf ( c int ) INHERITS (utility_inh_subroot);
CREATE

-- should complain as we are trying copy on a root or subroot
1U: COPY utility_inh_root FROM '/dev/null';
ERROR:  COPY FROM in utility mode is not supported on roots of inheritance and partition hierarchies
HINT:  consider using COPY FROM with the ON SEGMENT clause
1U: COPY utility_inh_subroot FROM '/dev/null';
ERROR:  COPY FROM in utility mode is not supported on roots of inheritance and partition hierarchies
HINT:  consider using COPY FROM with the ON SEGMENT clause

-- should be fine as we are trying it on a leaf
1U: COPY utility_inh_leaf FROM '/dev/null';
COPY 0

-- should complain as we are trying insert on a root or subroot
1U: INSERT INTO utility_inh_root VALUES(1);
ERROR:  INSERT in utility mode is not supported on roots of inheritance and partition hierarchies
1U: INSERT INTO utility_inh_subroot VALUES(1, 2);
ERROR:  INSERT in utility mode is not supported on roots of inheritance and partition hierarchies

-- should be fine as we are trying it on a leaf
1U: INSERT INTO utility_inh_leaf VALUES(1, 2, 3);
INSERT 1
